tests/components/powerwall/test_powerwall_data_manager.py

```diff
@@ -0,0 +1,51 @@
+"""Tests for the PowerwalLDataManager."""
+
+from unittest.mock import MagicMock, patch
+
+import pytest
+from tesla_powerwall import AccessDeniedError, Powerwall
+
+from homeassistant.components.powerwall import PowerwallDataManager
+from homeassistant.components.powerwall.models import PowerwallRuntimeData
+from homeassistant.exceptions import ConfigEntryAuthFailed
+
+
+def test_update_data_reauthenticate_on_access_denied(hass):
+    """Test of _update_data of PowerwallDataManager reauthenticates on AccessDeniedError."""
+    runtime_data = PowerwallRuntimeData(
```

Comment:
We aren't allowed to access integration internals in tests

https://developers.home-assistant.io/docs/development_testing?_highlight=testing#writing-tests-for-integrations

We need to patch the libraries and set the integration to simulate the auth fail to assert how it behaves externally

Reply:
I updated the tests to test the integration and not the internals, but I wasn't sure on how to correctly trigger the update. I am now using the following to get the coordinator and than update:
```python
await hass.data[DOMAIN][config_entry.entry_id][POWERWALL_COORDINATOR]._async_update_data()
```